pil-analyzer/src/traits_resolver.rs

```diff
@@ -1,131 +1,85 @@
 use powdr_ast::{
-    analyzed::{
-        Expression, FunctionValueDefinition, Identity, PolynomialReference, Reference, Symbol,
-        TypedExpression,
-    },
+    analyzed::{Expression, PolynomialReference},
     parsed::{
         types::{TupleType, Type},
-        visitor::AllChildren,
-        SelectedExpressions, TraitImplementation,
+        TraitImplementation,
     },
 };
 use std::{collections::HashMap, sync::Arc};
 
-type SolvedImpl = ((String, Vec<Type>), Arc<Expression>);
+use crate::type_unifier::Unifier;
 
-pub struct TraitsResolver<'a> {
-    definitions: &'a HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
-    identities: &'a Vec<Identity<SelectedExpressions<Expression>>>,
+/// Resolves trait implementations for all definitions and identities.
+/// Checks references for calls to trait implementations
+/// and unifies them using their type args.
+///
+/// Returns a nested map of trait name to a map of type args to the expression.
+pub fn traits_resolution(
+    references: Vec<PolynomialReference>,
+    trait_impls: &HashMap<String, Vec<TraitImplementation<Expression>>>,
+) -> HashMap<String, HashMap<Vec<Type>, Arc<Expression>>> {
+    TraitsResolver::new().resolve_traits(references, trait_impls)
 }
 
-impl<'a> TraitsResolver<'a> {
-    pub fn new(
-        definitions: &'a HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
-        identities: &'a Vec<Identity<SelectedExpressions<Expression>>>,
-    ) -> Self {
-        Self {
-            definitions,
-            identities,
-        }
+type SolvedImpl = ((String, Vec<Type>), Arc<Expression>);
+
+struct TraitsResolver {}
```

Comment:
Again, this is my taste, but maybe it would be nice to store the solved_impl map and the `trait_impls` in the struct and then call a function on this struct on a single reference at a time and then have a function to extract the solved_impl map (taking `self` by value). This is a rather flexible interface and we can then iterate over all expressions in the pil_analyzer. Also it is easier to handle the errors singe we can just return a simple Result here (because resolving the reference either fails or not).